### ISSUE TITLE
Using Upload service within an Android Service (bundled as cordova plugin)

### ISSUE THREAD

**Issue by azharbeebeejaun** [gotev/android-upload-service#580]
**What do you want to achieve?**
We are bundling it inside a cordova plugin. We have other wrapper logic which will handle the uploads and connection with a cordova app.
We'd like to start and use this library within an Android service. Can the UploadServiceConfig.initialize and createNotificationChannel etc be done on the onStartCommand of android.app.Service? Or would it be problematic for us, as we have seen on some devices (see Additional info)?

**What have you tried to do?**
We've looked at https://github.com/gotev/android-upload-service/wiki/Getting-Started-with-4.x and the 4.x docs.
It shows how to use android-upload-service starting from an Application class. We wish to use it from an Android Service.

**Which Upload Service Version are you using?**
4.0.0

**On which Android version?**
Android 9/10/11

**Additional info**
We were wrongly using this library because instead of being in the Application class, the UploadServiceConfig.initialize was being done later in the app. Which is why we saw some crashes like:
```
java.lang.RuntimeException: 
  at android.app.ActivityThread.handleCreateService (ActivityThread.java:3793)
  ...
Caused by: java.lang.IllegalArgumentException: 
  at net.gotev.uploadservice.UploadServiceConfig.getNamespace (UploadServiceConfig.kt:74)
  ...
```
This could be fixed by following the https://github.com/gotev/android-upload-service/wiki/Getting-Started-with-4.x wiki page.

The Service we're trying to achieve is available at https://github.com/spoonconsulting/cordova-plugin-background-upload/blob/master/src/android/ManagerService.java.

The fix we're thinking of to correctly do UploadServiceConfig.initialize is at https://github.com/spoonconsulting/cordova-plugin-background-upload/pull/137/files. But the idea is to switch completely to being service based since the cordova plugin aims to handle uploads even while offline.

Any help, comment or advice would be greatly appreciated. Thank you.

**Comment by gotev** [comment]
Thank you for getting in touch and the detailed explanation! 

I'll start by saying I have no Cordova experience as I work fully native on Android and iOS, but I can give you some advice about the native code of your plug-in.

Devs from Vydia have made a React Native wrapper of this library. Maybe you could get some insights also from their approach.

I've seen the code and from what I understood at a first glance you need to make uploads only when the network is available and you have all the bridge between JS and native code.

If you use Moshi instead of bundled JSONObject which is really bad and bugged, you will add in another dependency but you will gain more polished code and improved runtime performance.

I can't see the benefit of wrapping everything inside a service. You will have a lot of headaches as it over complicates your code and on Android 8 onwards it won't run in the background unless you add a foreground notification, which means having your wrapping service notification + Upload Service notification. Not a great UX and source of many lifecycle issues.

You can achieve all the same stuff also with a simple class which acts as a wrapper between the two worlds without a Service at all. Be sure to instantiate the class only once and to keep it around as long as the application lives. In the class constructor you can put the initialisation stuff you need and then you can use both ReactiveNetwork and UploadService inside, which are both already non blocking, so you don't need further wrapping to ensure not locking the main thread. Your code will be a lot simpler, too. You should call the initialisation where you init your App.

Beware of stuff like setMainActivity. As a rule of thumb, never retain context of any kind (activity, fragment, service, application) as this causes huge memory leaks and hardly debuggable issues. For example a simple device rotation recreates an activity context. That also happens when an app goes and then comes back from the background after a simple app switch. Use context only by injecting it where it's needed every time and never cache it. Caching context = memory leak. Tools like LeakCanary can help you spot a lot of that issues.

Also seen some defensive code to prevent too many upload updates in onProgress. Check out the progress interval configuration. Upload Service already has an internal throttling mechanism to prevent posting too many events. By default it sends Max of 6 events per second. Details in the Wiki.

You can also leverage UploadService's persistent uploads to be able to store requests and have them run also when your app or the device reboots. You can find details in the Wiki.

That's it for now! Let me know if you want to get your wrapper listed in the readme once it's finished 🥳

**Comment by azharbeebeejaun** [comment]
Thank you very much for taking the time to view the code and providing us with these insights. We are currently learning as we're doing and vice versa without really an experienced Android developer on the team. We'll certainly look at the items you raised as we progress.

The reason we went for the Service is because the use case is a field user going to different sites taking pictures without having an active connection. So, a foreground notification lives and waits for the connection to complete the upload/s. The field technician will be a business user and might not open the app again for the day so the OS will most certainly kill the app's process. We managed to get such a foreground service running based on your AbstractSingleNotificationHandler (subclass: [NotificationHandler.java](https://github.com/spoonconsulting/cordova-plugin-background-upload/blob/master/src/android/NotificationHandler.java)). We're not sure now whether we're setting ourselves for more crashes especially from UploadServiceConfig.initialize. Your comments have already directed us to the other places we can expect poor performance or crashes, thanks again!

**Comment by gotev** [comment]
Glad it's been useful!

**Comment by azharbeebeejaun** [comment]
Hi again,

Sorry to revive a closed thread. I had a question/feature request regarding the UploadService service. 
The service is started with [`START_STICKY`](https://github.com/gotev/android-upload-service/blob/master/uploadservice/src/main/java/net/gotev/uploadservice/UploadService.kt#L240) to be restarted in case the system kills it while it has uploads pending. 
Could this be configurable through the `taskCreationParameters` or something else? Would you be accepting a PR on this change?
We are currently using the library while the main app is in background or has already been killed by the OS. Thus, when the OS respawns the Upload Service, the service crashes with an IllegalArgumentException (like the original issue description) since the namespace is not present.

Regards,
Azhar

**Comment by gotev** [comment]
Is this a behavior you could replicate using the [demo app](https://github.com/gotev/android-upload-service/releases/latest)? If yes, open a new issue with all the details and we'll discuss it there 😉

**Comment by azharbeebeejaun** [comment]
I've tried replicating it on the demo app, but in vain. The namespace is persisted even if I force-stop and start the service again. It's different on a Cordova environment though. I'll dig deeper to see what solution I can find. Thank you.

**Comment by gotev** [comment]
In the native Android world, every instance you create during app execution is bound to the application context, which gets created when you tap on an app to open it. When you kill an app which has a running Service with START_STICKY, the system re-starts it again, recreating the application context. By initializing Android Upload Service in the Application's `onCreate` you're sure to meet system lifecycles properly.

I find it correct that you're unable to replicate the behavior with the demo app. I asked you to try to be sure your specific case was covered in the library implementation and guidelines. You can safely restrict your debugging to how the Cordova lifecycles works on Android.
